static/js/apps/eval_retrieval_generation/feedback_navigation.tsx

```diff
@@ -110,104 +198,102 @@ export function FeedbackNavigation(
     }
   };
 
-  // Button Conditions
-  const showPrevQuery = (): boolean => {
-    if (_.isEmpty(sortedQueryIds)) {
-      return false;
+  const nextEvalStage = async () => {
+    if (await props.checkAndSubmit()) {
+      setFeedbackStage(feedbackStageList[currStageIdx + 1]);
+      setSessionCallId(NEW_QUERY_CALL_ID);
     }
-    return (
-      sessionQueryId > sortedQueryIds[0] &&
-      feedbackStage === FeedbackStage.OVERALL
-    );
   };
 
-  const showPrev = (): boolean => {
-    return feedbackStage === FeedbackStage.CALLS;
-  };
+  // Button Conditions
 
-  const showNext = (): boolean => {
-    if (props.showNextOverride !== undefined) {
-      return props.showNextOverride;
+  function getNextButtonType(): ButtonType {
+    if (isEndOfStage(feedbackStage, sessionCallId, numCalls)) {
+      if (currStageIdx < feedbackStageList.length - 1) {
+        return ButtonType.NEXT_EVAL_STAGE;
+      }
+      const hasMoreQueries =
+        !_.isEmpty(sortedQueryIds) &&
+        sessionQueryId < sortedQueryIds[sortedQueryIds.length - 1];
+      return hasMoreQueries ? ButtonType.NEXT_QUERY : ButtonType.FINISH;
+    } else {
+      return ButtonType.NEXT;
     }
-    return sessionCallId < numCalls();
-  };
+  }
 
-  const showNextQuery = (): boolean => {
-    if (_.isEmpty(sortedQueryIds) || showNext()) {
-      return false;
+  function getPrevButtonType(): ButtonType {
+    if (isStartOfStage(feedbackStage, sessionCallId)) {
+      if (currStageIdx > 0) {
+        return ButtonType.PREV_EVAL_STAGE;
+      }
+      const hasPrevQuery =
+        !_.isEmpty(sortedQueryIds) && sessionQueryId > sortedQueryIds[0];
+      return hasPrevQuery ? ButtonType.PREV_QUERY : ButtonType.EMPTY;
+    } else {
+      return ButtonType.PREV;
     }
-    return sessionQueryId < sortedQueryIds[sortedQueryIds.length - 1];
-  };
+  }
+
+  function getOnClickFunction(buttonType: ButtonType) {
+    switch (buttonType) {
+      case ButtonType.PREV_QUERY:
+        return prevQuery;
+      case ButtonType.PREV:
+        return prev;
+      case ButtonType.PREV_EVAL_STAGE:
+        return prevEvalStage;
+      case ButtonType.NEXT_QUERY:
+        return nextQuery;
+      case ButtonType.NEXT:
+        return next;
+      case ButtonType.NEXT_EVAL_STAGE:
+        return nextEvalStage;
+      case ButtonType.FINISH:
+        return props.checkAndSubmit;
```

Comment:
default throw error?

Reply:
I don't think this would be an error, but instead added a default case with noop